cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/output/MultiOutputCommitter.java

```diff
@@ -122,8 +110,57 @@ public boolean isRecoverySupported() {
 
   @Override
   public void recoverTask(TaskAttemptContext taskContext) throws IOException {
+    delegateInParallel((name, delegate) -> {
+      TaskAttemptContext namedContext = MultiOutputFormat.getNamedTaskContext(taskContext, name);
+      delegate.recoverTask(namedContext);
+    });
+  }
+
+  private void delegateInParallel(DelegateFunction delegateFunction) throws IOException {
+    ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("multi-output-committer-%d").build();
+    ExecutorService executorService = Executors.newFixedThreadPool(delegates.size(), threadFactory);
```

Comment:
I think it might be better to bound the size to some upper limit, in case the number of delegates is too high, it will fail to create new thread.